src/postamble.js

```diff
@@ -175,7 +175,11 @@ Module['callMain'] = Module.callMain = function callMain(args) {
 #endif
       return;
     } else {
-      if (e && typeof e === 'object' && e.stack) Module.printErr('exception thrown: ' + [e, e.stack]);
+      var hasStack = e && typeof e === 'object' && e.stack;
+      if (hasStack) Module.printErr('exception thrown: ' + [e, e.stack]);
+      // In a shell environment, callMain is ran from a promise context. Exceptions thrown from
+      // promises don't set the shell's exit code, so call quit here to force it to be non-zero.
+      quitIfShell(1, hasStack ? '' : ('exception thrown: ' + e));
```

Comment:
I'm not sure if that makes sense. The way I read the `exit` function, it looks like it explicitly ends the current execution of `main`.

The other cases I know of where exceptions get thrown from wasm code is when calling JS imports that throw, and when a trap happens. In my understanding, we don't generally want to run more Module code when that happens.